File: 2007/4/14-4.py
#!/usr/bin/env python

import sys

def needed(x, p):
    x.sort()
    s = sum(x)
    mx = x[2]
    mn = x[0]
    md = x[1]
    if s - mx > p:
        return None
    elif mx + mn > p:
        return 2 * (mx + mn)
    elif 2 * (mx + mn) > p:
        return 2 * (md + mn)
    else:
        return mn + md

paper = int(sys.stdin.readline().split()[2])
sys.stdin.readline()

for i,l in enumerate(sys.stdin.read().splitlines()):
    print "Gift %s" % (i+1),
    d = [int(x) for x in l.split('x')]
    n = needed(d, paper)
    if n is None:
        print "cannot be wrapped."
    else:
        print "needs %s inches." % n

File: 2007/4/14-4.rb
#!/usr/bin/env ruby

def needed(x, p)
  s = x.inject(:+)
  mn, md, mx = x.sort
  if s - mx > p
    nil
  elsif mx + mn > p
    2 * (mx + mn)
  elsif 2 * (mx + mn) > p
    2 * (md + mn)
  else
    mn + md
  end
end

paper = STDIN.readline.split[2].to_i

STDIN.readline.to_i.times do |i|
  line = STDIN.readline
  print "Gift #{i+1} "
  d = line.split('x').collect { |a| a.to_i }
  n = needed(d, paper)
  if n
    puts "needs #{n} inches."
  else
    puts "cannot be wrapped."
  end
end
